pkg/reconcile/reconcile.go

```diff
@@ -0,0 +1,129 @@
+package reconcile
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	rbacv1 "k8s.io/api/rbac/v1"
+	apierror "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// DiffFunc takes two Kubernetes resources: expected and existing. Both are assumed to be
+// the same Kind. It compares the two, and returns an Operation indicating how to
+// transition from existing to expected. If an update is required, it will set the
+// relevant fields on existing to their intended values. This is so that we can simply
+// resubmit the existing resource, and any fields automatically set by the Kubernetes API
+// server will be retained.
+type DiffFunc func(runtime.Object, runtime.Object) Operation
+
+// Operation describes the operation performed by CreateOrUpdate.
+type Operation string
+
+var (
+	Create   Operation = "create"
+	Recreate Operation = "recreate"
+	Update   Operation = "update"
+	None     Operation = "none"
+	Error    Operation = "error"
+)
+
+// ObjWithMeta describes a Kubernetes resource with a metadata field. It's a combination
+// of two common existing Kubernetes interfaces. We do this because we want to use methods
+// from each in CreateOrUpdate, whilst still keeping the argument type generic.
+type ObjWithMeta interface {
```

Comment:
Any reason this is public? Do we expect anything outside of this package to make use of this interface?